include/swift/IDE/CodeCompletionResultType.h

```diff
@@ -107,54 +109,248 @@ enum class CodeCompletionResultTypeRelation : uint8_t {
+/// \c Int metatype and an \c Int instance type. Thus we consider both the
+/// instance and the metatype as result types of the expression \c Int.
 class CodeCompletionResultType {
-public:
-  enum class Flags : unsigned {
-    IsNotApplicable = 1 << 0,
-    /// If \p AlsoConsiderMetatype is set the code completion item will be
-    /// considered as producing the declared interface type (which is passed as
-    /// \p ResultTypes ) as well as the corresponding metatype.
-    /// This allows us to suggest 'Int' as 'Identical' for both of the following
-    /// functions
-    ///
-    ///   func receiveInstance(_: Int) {}
-    ///   func receiveMetatype(_: Int.Type) {}
-    AlsoConsiderMetatype = 1 << 1
-  };
+  /// \c ResultType1AndIsApplicable and \c ResultType2 store the following
+  /// information:
+  ///  - Whether the type is not applicable (see above). In this case
+  ///    \c ResultType1 and \c ResultType2 should be \c nullptr.
+  ///  - \c ResultType1 and \c ResultType2 form a collection of at most two
+  ///    result types. A \c CodeCompletionResultType cannot have a single unique
+  ///    type because for code completion we consider e.g. \c Int as producing
+  ///    both an \c Int metatype and an \c Int instance.
+  ///    It would be nice if we could use a SmallVector to store those types
+  ///    instead but \c CodeCompletionResultType is allocated in a bump
+  ///    allocator and freeing the bump allocator does not call the
+  ///    SmallVecotor's destructor, leaking potential heap memory. Since we only
+  ///    need two result types at the moment, I decided to implement it in a
+  ///    hacky way with two pointers.
+  /// The \c getResultTypes and \c isApplicable methods mask away this
+  /// implementation detail.
+  llvm::PointerIntPair<PointerUnion<Type, const USRBasedType *>, 1, bool>
+      ResultType1AndIsApplicable;
+  PointerUnion<Type, const USRBasedType *> ResultType2;
 
-private:
-  llvm::PointerIntPair<Type, 2, OptionSet<Flags>> TypeAndFlags;
+  llvm::SmallVector<PointerUnion<Type, const USRBasedType *>, 1>
+  getResultTypes() const {
+    if (ResultType1AndIsApplicable.getPointer() && ResultType2) {
+      return {ResultType1AndIsApplicable.getPointer(), ResultType2};
+    } else if (ResultType1AndIsApplicable.getPointer()) {
+      return {ResultType1AndIsApplicable.getPointer()};
+    } else {
+      assert(
+          !ResultType2 &&
+          "We shouldn't have a second result type if there was no first one");
+      return {};
+    }
+  }
 
-  CodeCompletionResultType(Type Ty, OptionSet<Flags> Flag)
-      : TypeAndFlags(Ty, Flag) {}
+  /// Memberwise initializer
+  CodeCompletionResultType(bool IsApplicable,
+                           PointerUnion<Type, const USRBasedType *> ResultType1,
+                           PointerUnion<Type, const USRBasedType *> ResultType2)
+      : ResultType1AndIsApplicable(ResultType1, IsApplicable),
+        ResultType2(ResultType2) {}
```

Comment:
Insert `assert(ResultType1.is<Type>() == ResultType2.is<Type>());` or do we have scenario mixing them?

Reply:
We don’t have a scenario where we’re mixing them but we never make assumptions that both of them are of the same kind either. Adding the assertion just adds an artificial restriction that’s not necessary IMO.